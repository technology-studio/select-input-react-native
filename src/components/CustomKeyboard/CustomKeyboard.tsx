import React, { Component } from 'react'
import {
  Dimensions, Modal, type StyleProp, TouchableWithoutFeedback, View,
} from 'react-native'

import { KeyboardButton } from '../KeyboardButton'

import { styles } from './styles'

type Props = {
  buttonsTextStyle?: StyleProp<Record<string, unknown>>,
  buttonsViewStyle?: StyleProp<Record<string, unknown>>,
  cancelKeyText: string,
  onCancelPress: () => void,
  onSubmitPress: () => void,
  submitKeyText: string,
  visible: boolean,
  children: JSX.Element | null,
}

export class CustomKeyboard extends Component<Props> {
  state = {
    width: Dimensions.get('window').width,
  }

  updateDimensions = (): void => {
    this.setState({
      width: Dimensions.get('window').width,
    })
  }

  onCancelPress = (): void => {
    const { onCancelPress } = this.props
    onCancelPress()
  }

  onSubmitPress = (): void => {
    const { onSubmitPress } = this.props
    onSubmitPress()
  }

  render (): JSX.Element {
    const { width } = this.state
    const {
      buttonsTextStyle,
      buttonsViewStyle,
      visible,
      cancelKeyText,
      submitKeyText,
      children,
    } = this.props

    return (
      <Modal
        animationType={'slide'}
        transparent={true}
        visible={visible}
        onOrientationChange={this.updateDimensions}
        supportedOrientations={[
          'portrait',
          'landscape',
          'landscape-left',
          'landscape-right',
        ]}
      >
        <TouchableWithoutFeedback>
          <View style={styles.container}>
            <View style={[styles.modal, { width }]}>
              <View style={[styles.buttonview, buttonsViewStyle, { width }]}>
                <KeyboardButton
                  onPress={this.onCancelPress}
                  text={cancelKeyText}
                  textAlign={'left'}
                  textStyle={buttonsTextStyle}
                />

                <KeyboardButton
                  onPress={this.onSubmitPress}
                  text={submitKeyText}
                  textAlign={'right'}
                  textStyle={buttonsTextStyle}
                />
              </View>

              <View>{children}</View>
            </View>
          </View>
        </TouchableWithoutFeedback>
      </Modal>
    )
  }
}
