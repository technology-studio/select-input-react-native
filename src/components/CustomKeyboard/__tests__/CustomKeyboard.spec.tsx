import React from 'react'
import renderer from 'react-test-renderer'

import CustomKeyboard from '../CustomKeyboard'

const props = {
  buttonsViewStyle: {
    backgroundColor: '#AAAAAA',
    borderColor: '#BBBBBB',
    borderBottomWidth: 1,
  },
  buttonsTextStyle: {
    color: '#CCCCCC',
    fontSize: 13,
  },
  cancelKeyText: 'cancelKeyText',
  onCancelPress: jest.fn(),
  onSubmitPress: jest.fn(),
  submitKeyText: 'submitKeyText',
  visible: false,
  children: null,
}

describe('CustomKeyboard', () => {
  describe('rendering', () => {
    // TODO: resolve issue with missing views in snapshot testing
    // test('renders correctly with given props', () => {
    //   const customKeyboard = renderer.create(<CustomKeyboard {...props} />)
    //   expect(customKeyboard).toMatchSnapshot()
    // })
  })

  describe('callbacks', () => {
    const onCancelPress = jest.fn()
    const onSubmitPress = jest.fn()

    const customKeyboard = renderer
      .create(
        <CustomKeyboard
          {...props}
          onCancelPress={onCancelPress}
          onSubmitPress={onSubmitPress}
        />,
      )
      .getInstance()

    test('call onCancelPress prop correctly', () => {
      // @ts-expect-error -- TODO: fix type
      customKeyboard?.onCancelPress()
      expect(onCancelPress).toHaveBeenCalled()
    })

    test('call onSubmitPress prop correctly', () => {
      // @ts-expect-error -- TODO: fix type
      customKeyboard?.onSubmitPress()
      expect(onSubmitPress).toHaveBeenCalled()
    })
  })
})
