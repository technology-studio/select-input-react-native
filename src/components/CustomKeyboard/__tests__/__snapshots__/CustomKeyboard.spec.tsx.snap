// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomKeyboard rendering renders correctly with given props 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  onOrientationChange={[Function]}
  supportedOrientations={
    Array [
      "portrait",
      "landscape",
      "landscape-left",
      "landscape-right",
    ]
  }
  transparent={true}
  visible={false}
>
  <View
    accessible={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "flex-end",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#F5FCFF",
            "justifyContent": "center",
            "padding": 0,
          },
          Object {
            "width": 750,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#CCCFD6",
              "borderBottomWidth": 0,
              "borderColor": "#CCCFD6",
              "borderTopColor": "lightgrey",
              "borderTopWidth": 0.5,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "padding": 0,
            },
            Object {
              "backgroundColor": "#AAAAAA",
              "borderBottomWidth": 1,
              "borderColor": "#BBBBBB",
            },
            Object {
              "width": 750,
            },
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#006BFF",
                  "padding": 8,
                },
                Object {
                  "color": "#CCCCCC",
                  "fontSize": 13,
                },
                Object {
                  "textAlign": "left",
                },
              ]
            }
          >
            cancelKeyText
          </Text>
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#006BFF",
                  "padding": 8,
                },
                Object {
                  "color": "#CCCCCC",
                  "fontSize": 13,
                },
                Object {
                  "textAlign": "right",
                },
              ]
            }
          >
            submitKeyText
          </Text>
        </View>
      </View>
      <View />
    </View>
  </View>
</Modal>
`;
